segment_store/src/segment.rs

```diff
@@ -335,65 +338,213 @@ impl Segment {
+                    (column_name, *typ)
+                })
+                .collect::<Vec<_>>(),
+            ..ReadGroupResult::default()
+        };
+
         // Handle case where there are no predicates and all the columns being
         // grouped support constant-time expression of the row_ids belonging to
         // each grouped value.
-        let all_group_cols_pre_computed = group_columns.iter().all(|col_name| {
-            self.column_by_name(col_name)
+        let all_group_cols_pre_computed = result.group_columns.iter().all(|name| {
+            self.column_by_name(name)
                 .properties()
                 .has_pre_computed_row_ids
         });
         if predicates.is_empty() && all_group_cols_pre_computed {
-            return self.read_group_all_rows_all_rle(group_columns, aggregates);
+            self.read_group_all_rows_all_rle(&mut result);
+            return result;
+        }
+
+        // If there is a single group column then we can use an optimised
+        // approach for building group keys
+        if group_columns.len() == 1 {
+            self.read_group_single_group_column(predicates, &mut result);
+            return result;
         }
 
-        ReadGroupResult::default()
+        // Perform the group by using a hashmap
+        self.read_group_hash(predicates, &mut result);
+        result
+    }
+
+    fn read_group_hash<'a>(&'a self, predicates: &[Predicate<'_>], dst: &mut ReadGroupResult<'a>) {
+        let row_ids = self.row_ids_from_predicates(predicates);
+        let filter_row_ids = match row_ids {
+            RowIDsOption::None(_) => return, // no matching rows
+            RowIDsOption::Some(row_ids) => Some(row_ids.to_vec()),
+            RowIDsOption::All(row_ids) => None,
+        };
+
+        let group_cols_num = dst.group_columns.len();
+        let agg_cols_num = dst.aggregate_columns.len();
+
+        // materialise all *encoded* values for each column we are grouping on.
+        // These will not be the logical (typically string) values, but will be
+        // vectors of integers representing the physical values.
+        let mut groupby_encoded_ids = Vec::with_capacity(group_cols_num);
+        for name in &dst.group_columns {
+            let col = self.column_by_name(name);
+            let mut encoded_values_buf = EncodedValues::with_capacity_u32(col.num_rows() as usize);
+
+            // do we want some rows for the column or all of them?
+            match &filter_row_ids {
+                Some(row_ids) => {
+                    encoded_values_buf = col.encoded_values(row_ids, encoded_values_buf);
+                }
+                None => {
+                    // None implies "no partial set of row ids" meaning get all of them.
+                    encoded_values_buf = col.all_encoded_values(encoded_values_buf);
+                }
+            }
+            groupby_encoded_ids.push(encoded_values_buf);
+        }
```

Comment:
Personally I would write this and other cases that have `Vec::with_capacity` followed by a `for` loop as:

```suggestion
        let groupby_encoded_ids: Vec<_> = dst.group_columns.iter().map(|name| {
            let col = self.column_by_name(name);
            let mut encoded_values_buf = EncodedValues::with_capacity_u32(col.num_rows() as usize);

            // do we want some rows for the column or all of them?
            match &filter_row_ids {
                Some(row_ids) => {
                    encoded_values_buf = col.encoded_values(row_ids, encoded_values_buf);
                }
                None => {
                    // None implies "no partial set of row ids" meaning get all of them.
                    encoded_values_buf = col.all_encoded_values(encoded_values_buf);
                }
            }
            encoded_values_buf
        }).collect();
```

I'm not sure if there are performance implications

Reply:
This looks good. No perf implications  really because it will likely only cost at most an extra vector growth on `groupby_encoded_ids`, but usually won't even cost that.